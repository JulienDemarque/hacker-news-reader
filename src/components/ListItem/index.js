import React from 'react';
import PropTypes from 'prop-types';
import Timeago from 'timeago.js';
import getSiteHostname from 'utils/getSiteHostname';

import { Item, Title, Host, ExernalLink, Description, CommentLink } from './styles';

const timeago = Timeago();

const HN_ROOT = 'https://news.ycombinator.com';
const HN_ITEM = `${HN_ROOT}/item?id=`;
const HN_USER = `${HN_ROOT}/user?id=`;

const ListItem = ({ by, kids = [], score, url, title, id, type, time }) => {
  const commentUrl = `${HN_ITEM}${id}`;
  const link = type === 'story' ? commentUrl : url;
  const site = getSiteHostname(url) || 'news.ycombinator.com';

  return (
    <Item>
      <ExernalLink href={link} rel="nofollow noreferrer noopener" target="_blank">
        <Title>
          {title} <Host>({site})</Host>
        </Title>
      </ExernalLink>
      <Description>
        {score} points by{' '}
        <CommentLink href={`${HN_USER}${by}`} rel="nofollow noreferrer noopener" target="_blank">
          {by}
        </CommentLink>{' '}
        {timeago.format(new Date(time * 1000).toISOString())} {' | '}
        <CommentLink href={commentUrl} rel="nofollow noreferrer noopener" target="_blank">
          {kids.length} Comments
        </CommentLink>
      </Description>
    </Item>
  );
};

ListItem.propTypes = {
  by: PropTypes.string.isRequired,
  kids: PropTypes.array,
  score: PropTypes.number.isRequired,
  url: PropTypes.string,
  title: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  time: PropTypes.number.isRequired,
};

export default ListItem;
