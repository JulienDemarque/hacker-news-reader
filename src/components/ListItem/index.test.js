import React from 'react';

import { shallow } from 'enzyme';
import ListItem from './index';

it('renders without crashing', () => {
  const component = shallow(
    <ListItem
      by="NN88"
      kids={[
        18277581,
        18277336,
        18277409,
        18277173,
        18277196,
        18277523,
        18277390,
        18277302,
        18277324,
        18277514,
        18277389,
      ]}
      score={58}
      url="http://nymag.com/intelligencer/2018/10/retailers-are-using-facial-recognition-technology-too.html"
      title="The Secretive Business of Facial-Recognition Software in Retail Stores"
      id={18276862}
      type="story"
      time={1540231727}
    />,
  );

  expect(component).toMatchSnapshot();
});
