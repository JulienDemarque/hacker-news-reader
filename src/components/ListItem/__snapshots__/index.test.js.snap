// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListItem
    by="NN88"
    id={18276862}
    kids={
        Array [
            18277581,
            18277336,
            18277409,
            18277173,
            18277196,
            18277523,
            18277390,
            18277302,
            18277324,
            18277514,
            18277389,
          ]
    }
    score={58}
    time={1540231727}
    title="The Secretive Business of Facial-Recognition Software in Retail Stores"
    type="story"
    url="http://nymag.com/intelligencer/2018/10/retailers-are-using-facial-recognition-technology-too.html"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.a
          href="http://nymag.com/intelligencer/2018/10/retailers-are-using-facial-recognition-technology-too.html"
          rel="nofollow noreferrer noopener"
          target="_blank"
>
          <styled.h3>
                    The Secretive Business of Facial-Recognition Software in Retail Stores
                     
                    <styled.span>
                              (
                              nymag.com
                              )
                    </styled.span>
          </styled.h3>
</styled.a>,
        <styled.div>
          58
           points by
           
          <styled.a
                    href="https://news.ycombinator.com/user?id=NN88"
                    rel="nofollow noreferrer noopener"
                    target="_blank"
          >
                    NN88
          </styled.a>
           
          1 hour ago
           
           | 
          <styled.a
                    href="https://news.ycombinator.com/item?id=18276862"
                    rel="nofollow noreferrer noopener"
                    target="_blank"
          >
                    11
                     Comments
          </styled.a>
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <styled.h3>
            The Secretive Business of Facial-Recognition Software in Retail Stores
             
            <styled.span>
                        (
                        nymag.com
                        )
            </styled.span>
</styled.h3>,
          "href": "http://nymag.com/intelligencer/2018/10/retailers-are-using-facial-recognition-technology-too.html",
          "rel": "nofollow noreferrer noopener",
          "target": "_blank",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "The Secretive Business of Facial-Recognition Software in Retail Stores",
              " ",
              <styled.span>
                (
                nymag.com
                )
</styled.span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "The Secretive Business of Facial-Recognition Software in Retail Stores",
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "(",
                  "nymag.com",
                  ")",
                ],
              },
              "ref": null,
              "rendered": Array [
                "(",
                "nymag.com",
                ")",
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            58,
            " points by",
            " ",
            <styled.a
              href="https://news.ycombinator.com/user?id=NN88"
              rel="nofollow noreferrer noopener"
              target="_blank"
>
              NN88
</styled.a>,
            " ",
            "1 hour ago",
            " ",
            " | ",
            <styled.a
              href="https://news.ycombinator.com/item?id=18276862"
              rel="nofollow noreferrer noopener"
              target="_blank"
>
              11
               Comments
</styled.a>,
          ],
        },
        "ref": null,
        "rendered": Array [
          58,
          " points by",
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "NN88",
              "href": "https://news.ycombinator.com/user?id=NN88",
              "rel": "nofollow noreferrer noopener",
              "target": "_blank",
            },
            "ref": null,
            "rendered": "NN88",
            "type": [Function],
          },
          " ",
          "1 hour ago",
          " ",
          " | ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                11,
                " Comments",
              ],
              "href": "https://news.ycombinator.com/item?id=18276862",
              "rel": "nofollow noreferrer noopener",
              "target": "_blank",
            },
            "ref": null,
            "rendered": Array [
              11,
              " Comments",
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.a
            href="http://nymag.com/intelligencer/2018/10/retailers-are-using-facial-recognition-technology-too.html"
            rel="nofollow noreferrer noopener"
            target="_blank"
>
            <styled.h3>
                        The Secretive Business of Facial-Recognition Software in Retail Stores
                         
                        <styled.span>
                                    (
                                    nymag.com
                                    )
                        </styled.span>
            </styled.h3>
</styled.a>,
          <styled.div>
            58
             points by
             
            <styled.a
                        href="https://news.ycombinator.com/user?id=NN88"
                        rel="nofollow noreferrer noopener"
                        target="_blank"
            >
                        NN88
            </styled.a>
             
            1 hour ago
             
             | 
            <styled.a
                        href="https://news.ycombinator.com/item?id=18276862"
                        rel="nofollow noreferrer noopener"
                        target="_blank"
            >
                        11
                         Comments
            </styled.a>
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.h3>
              The Secretive Business of Facial-Recognition Software in Retail Stores
               
              <styled.span>
                            (
                            nymag.com
                            )
              </styled.span>
</styled.h3>,
            "href": "http://nymag.com/intelligencer/2018/10/retailers-are-using-facial-recognition-technology-too.html",
            "rel": "nofollow noreferrer noopener",
            "target": "_blank",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "The Secretive Business of Facial-Recognition Software in Retail Stores",
                " ",
                <styled.span>
                  (
                  nymag.com
                  )
</styled.span>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "The Secretive Business of Facial-Recognition Software in Retail Stores",
              " ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    "(",
                    "nymag.com",
                    ")",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "(",
                  "nymag.com",
                  ")",
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              58,
              " points by",
              " ",
              <styled.a
                href="https://news.ycombinator.com/user?id=NN88"
                rel="nofollow noreferrer noopener"
                target="_blank"
>
                NN88
</styled.a>,
              " ",
              "1 hour ago",
              " ",
              " | ",
              <styled.a
                href="https://news.ycombinator.com/item?id=18276862"
                rel="nofollow noreferrer noopener"
                target="_blank"
>
                11
                 Comments
</styled.a>,
            ],
          },
          "ref": null,
          "rendered": Array [
            58,
            " points by",
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "NN88",
                "href": "https://news.ycombinator.com/user?id=NN88",
                "rel": "nofollow noreferrer noopener",
                "target": "_blank",
              },
              "ref": null,
              "rendered": "NN88",
              "type": [Function],
            },
            " ",
            "1 hour ago",
            " ",
            " | ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  11,
                  " Comments",
                ],
                "href": "https://news.ycombinator.com/item?id=18276862",
                "rel": "nofollow noreferrer noopener",
                "target": "_blank",
              },
              "ref": null,
              "rendered": Array [
                11,
                " Comments",
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
